tycho-its/src/test/java/org/eclipse/tycho/test/reactor/makeBehaviour/MavenReactorMakeOptionsTest.java

```diff
@@ -0,0 +1,105 @@
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Test Maven reactor make behaviours
+ *
+ * Test project dependencies:
+ *
+ * <pre>
+ * feature2 -> feature1,bundle2
+ * feature1 -> bundle1
+ * </pre>
+ *
+ */
+public class MavenReactorMakeOptionsTest extends AbstractTychoIntegrationTest {
+
+	private Verifier verifier;
+
+	@Before
+	public void setUp() throws Exception {
+		verifier = getVerifier("reactor.makeBehaviour");
+		verifier.executeGoal("clean");
+	}
+
+	@Test
+	public void testCompleteBuild() throws Exception {
+		verifier.executeGoal("verify");
+		verifier.verifyErrorFreeLog();
+		verifier.assertFilePresent("bundle1/target/bundle1-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("bundle2/target/bundle2-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("feature1/target/feature1-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("feature2/target/feature2-1.0.0-SNAPSHOT.jar");
+	}
+
+	@Test
+	public void testAlsoMake() throws Exception {
+		// REACTOR_MAKE_UPSTREAM
+		verifier.addCliOption("-am");
+		verifier.addCliOption("-pl feature1");
+		verifier.executeGoal("verify");
+		verifier.verifyErrorFreeLog();
+		verifier.assertFilePresent("bundle1/target/bundle1-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("feature1/target/feature1-1.0.0-SNAPSHOT.jar");
+	}
+
+	@Test
+	public void testAlsoMakeDependents() throws Exception {
+		// REACTOR_MAKE_DOWNSTREAM
+		verifier.addCliOption("-amd");
+		verifier.addCliOption("-pl bundle1,bundle2");
+		verifier.executeGoal("verify");
+		verifier.verifyErrorFreeLog();
+		verifier.assertFilePresent("bundle1/target/bundle1-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("bundle2/target/bundle2-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("feature1/target/feature1-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("feature2/target/feature2-1.0.0-SNAPSHOT.jar");
+	}
+
+	@Test
+	public void testBoth() throws Exception {
+		// REACTOR_MAKE_BOTH
+		verifier.addCliOption("-am");
+		verifier.addCliOption("-amd");
+		verifier.addCliOption("-pl feature1,bundle2");
+		verifier.executeGoal("verify");
+		verifier.verifyErrorFreeLog();
+		verifier.assertFilePresent("bundle1/target/bundle1-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("bundle2/target/bundle2-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("feature1/target/feature1-1.0.0-SNAPSHOT.jar");
+		verifier.assertFilePresent("feature2/target/feature2-1.0.0-SNAPSHOT.jar");
+	}
+
+	@Test
+	public void testSingleProjectNoOptionFails() throws Exception {
+		try {
+			verifier.addCliOption("-pl feature1");
+			verifier.executeGoal("verify");
+			fail("Build should fail due to missing reactor dependency");
+		} catch (VerificationException e) {
+			verifier.verifyTextInLog("TODO complete");
+		}
+	}
+
+	@Test
+	public void testDownstreamFailsIfMissingDownstreamDependency() throws Exception {
+		// Downstream brings in feature2 but this requires bundle2 which is unrelated
```

Comment:
I have a question here, is this really desired? Is it not expected that dependencies of downstream are also included? Is this documented somewhere how it is supposed to behave?

Reply:
Eg my current implementation computes the following:

```
[INFO] [[ project feature1 depends on: ]]
[INFO]  IU: bundle1 1.0.0.qualifier [of project bundle1]
```


```
[INFO] Computing additional make-downstream dependencies based on initial project set of feature1, bundle1
[INFO]  + add downstream project bundle1 of feature1...
```

So how is feature2 supposed to be pulled in?

Reply:
I think because in the dependency DAG feature2 is not a strict dependent of bundle1,feature1 as bundle2 is on a different branch and not dependent on any of the projects in the list.

I agree that in reality I can't imagine a practical case where this would be useful behaviour.

I wrote this to match behaviour of a vanilla maven reactor build.